opt_adapt/opt.py

```diff
@@ -52,11 +53,33 @@ def __init__(self, options={}):
                 raise ValueError(f"Option {key} not recognised")
             self.__setattr__(key, value)
 
+# Only works for problem with control paramter in r-space
+def line_search(forward_run, mesh, u, P, J, dJ, Rspace, alpha=1e-4, max_search_iter=100):
+    
+    lr = 1
+    if Rspace:
+        initial_slope = float(dJ) * float(P)
+    else:
+        initial_slope = np.dot(dJ.dat.data, P.dat.data)
+
+    if initial_slope==0.0:
+        return 1.0
```

Comment:
Checking if something `==0.0` doesn't really work for floating point numbers. It is better to use `if np.isclose(initial_slope, 0.0)`.

Reply:
Again, it would be good to return `params.lr` as the default in this case.